torch/utils/data/dataloader.py

```diff
@@ -399,6 +403,90 @@ def __len__(self) -> int:
+        #
+        # eg. If current system has 2 physical CPUs with 16 cores each. And each core support 2
+        #     threads, then the total logical cpus here is 2 * 16 * 2 = 64. Let's say current
+        #     DataLoader process can use half of them which is 32, then the rational max number of
+        #     thread that initiated from this process is 32 * 2 = 64.
+        #     Now, let's say the created DataLoader has num_works = 4, and each worker is allowed
+        #     to create a max of 24 threads (get/set by torch.get/set_num_threads()), so the max
+        #     number of threads which can be created by this DataLoader process is 4 * 24 = 96
+        #     which is bigger than 64. So the warning message is triggered in this case.
+        #
+        #     To lower the max total number of therad that will be created by DataLoader, there
+        #     are 2 ways. One is to use torch.set_num_threads() to adjust the max thread number of
+        #     each worker, the other way is to make number of worker smaller.
+        #
+        # [Note] Please note that this function repects `cpuset` only when os.sched_getaffinity is
+        #        available (available in most of Linux system, but not OSX and Windows).
+        #        When os.sched_getaffinity is not available, os.cpu_count() is called instead, but
+        #        it doesn't repect cpuset.
+        def _create_warning_msg(max_td_created, max_td_per_worker, max_td_suggest, num_workers, cpuset_checked):
+
+            suggested_max_thread_msg = ((
+                "Our suggested max number of thread in current system is {}{}, which is smaller "
+                "than what this DataLoader might create.").format(
+                    max_td_suggest,
+                    ("" if cpuset_checked else " (`cpuset` is not taken into account)"))
+            ) if max_td_suggest is not None else (
+                "DataLoader is not able to compute a suggested max number of thread in current system.")
+
+            warn_msg = (
+                "This DataLoader might create a max of {} threads in total "
+                "(max number of thread in each worker {} multiply number of workers {}). {} "
+                "Please be aware that excessive threads might get DataLoader running slow or even freeze, "
+                "adjust the max number of thread for each worker through torch.set_num_threads(<num of thread>), "
+                "or adjust number of workers to avoid potential slowness/freeze if necessary.").format(
+                    max_td_created,
+                    max_td_per_worker,
+                    num_workers,
+                    suggested_max_thread_msg)
+            return warn_msg
+
+        if not self.num_workers or self.num_workers == 0:
+            return
+
+        max_num_thread_per_worker = torch.get_num_threads()
+        max_total_thread_created = max_num_thread_per_worker * self.num_workers
```

Comment:
I think `get_num_threads` is implemented here.
https://github.com/pytorch/pytorch/blob/b7e044f0e58624676130b2ef6a376a453fc6b090/aten/src/ATen/ParallelNative.cpp#L216-L234
It gives you different result from your expectation. In the `DataLoader`, it may not provide the maximum thread per worker.

Reply:
Yep, the implementation has multiple paths which result in multiple possibilities, I think normally, this api should work fine, since the main process and sub process normally has same env. But my problem currently is that whether we should check thread numbers given that we do limit thread number to 1 in the worker. I will dig this a bit, probably, we just need to suggest a worker number instead of taken threading into account.